
public class Program1136
{
    public static void main(String A[])
    {
        int iRet = 0;

        SinglyLL sobj = new SinglyLL();

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);
        sobj.InsertLast(101);
        sobj.InsertLast(111);

        sobj.Display();

        iRet = sobj.MiddleElementX();

        System.out.println("Middle element is : "+iRet);
    }
}

class node
{
    public int data;
    public node next;

    public node(int no)
    {
        this.data = no;
        this.next = null;
    }
}

class SinglyLL
{
    private node first;
    private int iCount;

    public SinglyLL()
    {
        this.first = null;
        this.iCount = 0;
    }

    public void InsertFirst(int iNo)
    {
        node newn = new node(iNo);

        newn.next = first;
        first = newn;

        iCount++;
    }

    public void InsertLast(int iNo)
    {
        node newn = new node(iNo);

        if(first == null)
        {
            first = newn;
        }
        else
        {
            node temp = first;

            while(temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
        }

        iCount++;
    }

    public int Count()
    {
        return iCount;
    }

    public void Display()
    {
        node temp = null;

        temp = first;

        while(temp != null)
        {
            System.out.print("| "+temp.data+" | -> ");
            temp = temp.next;
        }

        System.out.println("null");
    }

    // Time : N + N/2
    
    public int MiddleElement()
    {
        int iCount = Count();

        int Middle = iCount / 2;
        int i = 0;
        node temp = first;

        for(i = 1; i <= Middle; i++)
        {
            temp = temp.next;
        }

        return temp.data;
    }

    public int MiddleElementX()
    {
        node fast = first;
        node slow = first;

        while(fast != null) // Issue
        {
            fast = fast.next.next;
            slow = slow.next;
        }   

        return slow.data;
    }
}
